prov/lnx/src/lnx_domain.c

```diff
@@ -0,0 +1,584 @@
+					mem_desc->desc[i].prov->lpv_prov_name, rc);
+				frc = rc;
+			}
+		}
+	}
+
+	return frc;
+}
+
+static int lnx_mr_control(struct fid *fid, int command, void *arg)
+{
+	int i, rc, frc = 0;
+	struct fid_mr *mr, *cmr;
+	struct lnx_mem_desc *mem_desc;
+	struct lnx_mem_desc_prov *desc;
+
+	if (command != FI_ENABLE)
+		return -FI_ENOSYS;
+
+	mr = container_of(fid, struct fid_mr, fid);
+
+	mem_desc = mr->mem_desc;
+
+	/* TODO: This is another issue here because MR registration can happen
+	 * quiet often
+	 */
+	for (i = 0; i < mem_desc->desc_count; i++) {
+		desc = &mem_desc->desc[i];
+		cmr = desc->core_mr;
+		if (!cmr)
+			continue;
+		rc = fi_mr_enable(cmr);
+		if (rc) {
+			FI_WARN(&lnx_prov, FI_LOG_CORE, "%s lnx_mr_control() failed: %d\n",
+				mem_desc->desc[i].prov->lpv_prov_name, rc);
+			frc = rc;
+		}
+	}
+
+	return frc;
+}
+
+static struct fi_ops lnx_mr_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = lnx_mr_close,
+	.bind = lnx_mr_bind,
+	.control = lnx_mr_control,
+	.ops_open = fi_no_ops_open
+};
+
+static int
+lnx_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
+	       uint64_t flags, struct fid_mr **mr_fid)
+{
+	/*
+	 * If the address is specified then use it to find out which
+	 * domain to register the memory against. LNX can be managing
+	 * multiple underlying core provider endpoints, I need to register the
+	 * memory against the correct one.
+	 *
+	 * Once the domain is determined, I need to set the mr->mem_desc to
+	 * point to a structure which contains my local endpoint I'll end up
+	 * using (which is the same one that I registered the memory against)
+	 * and the associate fid_mr which the core provider set for me.
+	 *
+	 * I return that to the application.
+	 *
+	 * When the application calls back into the data operations API it'll
+	 * pass the mr. I can then pull out a pointer to my local endpoint
+	 * which I'll use in the data operation and pass it the correct mr.
+	 *
+	 * If the address is not provided, then I'll register the memory
+	 * buffer against all my core domains, store those and return them to
+	 * the user
+	 */
+
+	struct lnx_domain *domain;
+	struct lnx_fabric *fabric;
+	struct lnx_mr *lnx_mr = NULL;;
+	struct ofi_mr *mr;
+	struct lnx_mem_desc *mem_desc;
+	struct local_prov *entry;
+	int rc = 0, i = 1;
+	bool shm = false;
+
+	if (fid->fclass != FI_CLASS_DOMAIN || !attr || attr->iov_count <= 0)
+		return -FI_EINVAL;
+
+	domain = container_of(fid, struct lnx_domain, ld_domain.domain_fid.fid);
+	fabric = domain->ld_fabric;
+
+	lnx_mr = ofi_buf_alloc(fabric->mem_reg_bp);
+	if (!lnx_mr) {
+		rc = -FI_ENOMEM;
+		goto fail;
+	}
+
+	mr = &lnx_mr->mr;
+	mem_desc = &lnx_mr->desc;
+
+	mr->mr_fid.fid.fclass = FI_CLASS_MR;
+	mr->mr_fid.fid.context = attr->context;
+	mr->mr_fid.fid.ops = &lnx_mr_fi_ops;
+	mr->mr_fid.mem_desc = mem_desc;
+	mr->domain = &domain->ld_domain;
+	mr->flags = flags;
+
+	/* TODO: What's gonna happen if you try to register the same piece
```

Comment:
I think so. Just flagging it here because I have some concerns with the impact on performance and I'm not 100% sure if there are side effects of registering the same piece of memory with multiple providers at the same time. I don't think this has been done before and we need to be sure if it's not going to introduce issues going forward.